kafka-producer/src/main/java/com/expedia/www/haystack/pipes/kafkaProducer/KafkaToKafkaPipeline.java

```diff
@@ -0,0 +1,137 @@
+ *       See the License for the specific language governing permissions and
+ *       limitations under the License.
+ *
+ */
+package com.expedia.www.haystack.pipes.kafkaProducer;
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+import com.expedia.open.tracing.Span;
+import com.expedia.www.haystack.pipes.commons.kafka.TagFlattener;
+import com.expedia.www.haystack.pipes.kafkaProducer.config.KafkaProducerConfig;
+import com.expedia.www.haystack.pipes.key.extractor.SpanKeyExtractor;
+import com.netflix.servo.util.VisibleForTesting;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.pool2.ObjectPool;
+import org.apache.commons.pool2.impl.GenericObjectPool;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.streams.kstream.ForeachAction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class KafkaToKafkaPipeline implements ForeachAction<String, Span> {
+    @VisibleForTesting
+    static final String ERROR_MSG =
+            "Exception posting JSON [%s] to Kafka; received message [%s]";
+    static ObjectPool<KafkaCallback> OBJECT_POOL = new GenericObjectPool<>(
+            new CallbackFactory());
+    @VisibleForTesting
+    static Logger logger = LoggerFactory.getLogger(KafkaToKafkaPipeline.class);
+    @VisibleForTesting
+    static Factory factory = new KafkaToKafkaPipeline.Factory();
+    @VisibleForTesting
+    static Counter kafkaProducerCounter;
+    @VisibleForTesting
+    static Map<KafkaProducer<String, String>, List<SpanKeyExtractor>> kafkaProducerSpanExtractorMap = new HashMap<>();
+    private final TagFlattener tagFlattener = new TagFlattener();
+    private final Counter requestCounter;
+    private final Timer kafkaProducerTimer;
+    private List<KafkaProducerConfig> kafkaProducerConfigMaps;
+    private List<SpanKeyExtractor> spanKeyExtractors;
+
+    public KafkaToKafkaPipeline(MetricRegistry metricRegistry,
+                                ProjectConfiguration projectConfiguration,
+                                List<SpanKeyExtractor> spanKeyExtractors) {
+        this.kafkaProducerConfigMaps = projectConfiguration.getKafkaProducerConfigList();
+        this.spanKeyExtractors = spanKeyExtractors;
+        this.requestCounter = metricRegistry.counter("REQUEST");
+        this.kafkaProducerTimer = metricRegistry.timer("KAFKA_PRODUCER_POST_TIMER");
+        kafkaProducerCounter = metricRegistry.counter("KAFKA_PRODUCER_POST_COUNTER");
+    }
+
+    @Override
+    public void apply(String key, Span value) {
+        requestCounter.inc();
+        Timer.Context time = kafkaProducerTimer.time();
+        createKafkaProducersExtractorMap(kafkaProducerConfigMaps);
```

Comment:
We would never want to do this for each span. Should be done at the class initialisation and only once